pkg/controller/stash.go

```diff
@@ -218,23 +222,78 @@ func (c *StashController) getTotalHosts(target interface{}, namespace string) (*
-	case apis.KindStatefulSet:
-		ss, err := c.kubeClient.AppsV1().StatefulSets(namespace).Get(targetRef.Name, metav1.GetOptions{})
-		if err != nil {
-			return nil, err
+	if driver == api_v1beta1.VolumeSnapshotter {
+		switch targetRef.Kind {
+		case apis.KindStatefulSet:
+			ss, err := c.kubeClient.AppsV1().StatefulSets(namespace).Get(targetRef.Name, metav1.GetOptions{})
+			if err != nil {
+				return nil, err
+			}
+			return types.Int32P(types.Int32(ss.Spec.Replicas) * int32(len(ss.Spec.VolumeClaimTemplates))), err
+		case apis.KindDeployment:
+			deployment, err := c.kubeClient.AppsV1().Deployments(namespace).Get(targetRef.Name, metav1.GetOptions{})
+			if err != nil {
+				return nil, err
+			}
+			vollist := deployment.Spec.Template.Spec.Volumes
+			return countPVC(vollist), err
+
+		case apis.KindDaemonSet:
+			daemon, err := c.kubeClient.AppsV1().DaemonSets(namespace).Get(targetRef.Name, metav1.GetOptions{})
+			if err != nil {
+				return nil, err
+			}
+			vollist := daemon.Spec.Template.Spec.Volumes
+			return countPVC(vollist), err
+
+		case apis.KindReplicaSet:
+			RS, err := c.kubeClient.AppsV1().StatefulSets(namespace).Get(targetRef.Name, metav1.GetOptions{})
+			if err != nil {
+				return nil, err
+			}
+			vollist := RS.Spec.Template.Spec.Volumes
+			return countPVC(vollist), err
+
+		case apis.KindReplicationController:
+			RC, err := c.kubeClient.CoreV1().ReplicationControllers(namespace).Get(targetRef.Name, metav1.GetOptions{})
+			if err != nil {
+				return nil, err
+			}
+			vollist := RC.Spec.Template.Spec.Volumes
+			return countPVC(vollist), err
+
+		default:
+			return types.Int32P(1), nil
 		}
-		return ss.Spec.Replicas, nil
-	// all Daemon pod will take backup/restore. so total number of hosts will be number of ready replicas
-	case apis.KindDaemonSet:
-		dmn, err := c.kubeClient.AppsV1().DaemonSets(namespace).Get(targetRef.Name, metav1.GetOptions{})
-		if err != nil {
-			return nil, err
+	} else {
+		switch targetRef.Kind {
+		// all replicas of StatefulSet will take backup/restore. so total number of hosts will be number of replicas.
+		case apis.KindStatefulSet:
+			ss, err := c.kubeClient.AppsV1().StatefulSets(namespace).Get(targetRef.Name, metav1.GetOptions{})
+			if err != nil {
+				return nil, err
+			}
+			return ss.Spec.Replicas, nil
+		// all Daemon pod will take backup/restore. so total number of hosts will be number of ready replicas
+		case apis.KindDaemonSet:
+			dmn, err := c.kubeClient.AppsV1().DaemonSets(namespace).Get(targetRef.Name, metav1.GetOptions{})
+			if err != nil {
+				return nil, err
+			}
+			return &dmn.Status.NumberReady, nil
+		// for all other workloads, only one replica will take backup/restore. so number of total host will be 1
+		default:
+			return types.Int32P(1), nil
+		}
+	}
+}
+
+func countPVC(vollist []core.Volume) *int32 {
+	count = 0
+	for _, list := range vollist {
```

Comment:
rename iterator variable to `vol` intead of `list`. It's a a volume not a list of volume.